libsrc/c64/crt0.s

```diff
@@ -98,14 +85,51 @@ L2:     lda     zpsave,x
 
         rts
 
+
 ; ------------------------------------------------------------------------
-; Data
 
-.segment        "ZPSAVE"
+.segment        "INIT"
 
-zpsave: .res    zpspace
+initstart:
 
-.bss
+; Save the zero-page locations that we need.
+
+        ldx     #zpspace-1
+L1:     lda     sp,x
+        sta     zpsave,x
+        dex
+        bpl     L1
+
+; Clear the BSS data.
+
+        jsr     zerobss
+
+; Set up the stack.
+
+        lda     #<(__RAM_START__ + __RAM_SIZE__ + __STACKSIZE__)
+        ldx     #>(__RAM_START__ + __RAM_SIZE__ + __STACKSIZE__)
+        sta     sp
+        stx     sp+1            ; Set argument stack ptr
+
+; Call the module constructors.
+
+        jsr     initlib
+
+; Push the command-line arguments; and, call main().
+
+        jmp     callmain
+
+
+; ------------------------------------------------------------------------
+; Data
+
+.data
 
-spsave: .res    1
 mmusave:.res    1
+spsave: .res    1
```

Comment:
Just for clarification:

If we have a data structure that is partially initialized with something non-zero - and which therefore lives in the DATA segment - then it is of course an option to use `.res` to initialize the rest of the data structure with zeros. Otherwise you might have to write lots of `0` literals into the source code. An example:

```
.data
argv:   .addr   name
        .res    MAXARGS * 2
```

In the example below however we wouldn't have multiple `0` literals:

```
.data
ctrl:   .addr   _read
        .res    2                       ; CALLERDATA
        .res    2                       ; MODULE
        .res    2                       ; MODULE_SIZE
        .res    2                       ; MODULE_ID
```

I guess Uz and Greg see a parallel to the unnecessary explicit initialization of global (or static) C variables with zero (although that technically happens in the BSS segment). You see:

```
int my_global = 0;
```
vs.
```
int my_global;
```

In the same way the code above doesn't emphasize that `my_global` is initialized (although it is anyway) a `.res 2 my_global` doesn't emphasize that `my_global` is initialized (although it is anyway).

While I believe I understand the argument I don't follow it for two reasons:

* I always just write `int my_global;` even _IF_ the code depends on the implicit initialization - after all that's what it's for.

* The `.res` approach looses the benefits of 16-bit types. I consider the code below more readable:

```
.data
ctrl:   .addr   _read
        .addr   0                       ; CALLERDATA
        .addr   0                       ; MODULE
        .word   0                       ; MODULE_SIZE
        .word   0                       ; MODULE_ID
```